src/connection.rs

```diff
@@ -614,7 +614,13 @@ where
                         continue
                     }
                     inner.on_drop(Action::None);
-                    return Err(io::ErrorKind::WouldBlock.into())
+                    let can_read = inner.streams.get(&self.id).map(|s| s.state().can_read());
+                    if can_read.unwrap_or(false) {
+                        return Err(io::ErrorKind::WouldBlock.into())
+                    } else {
+                        debug!("{}: {}: can no longer read", inner.id, self.id);
+                        return Ok(0) // stream has been reset
```

Comment:
I would say that we should return an error if the stream has been reset, and EOF if it has been closed.

Reply:
Maybe this can be postponed? Within this PR we can not distinguish between close and reset and the check should only fix the problem described in the commit message. Once this and #37 are in we can focus on shutdown and FIN and then change this maybe.